### PR TITLE
[SV-COMP'18 13/19] Memcpy assertions

### PR DESCRIPTION
**Pull request by tautschnig**
Do not merge: this needs cleanup and review as well as regression tests.

### CODE REVIEW

File: src/ansi-c/ansi_c_internal_additions.cpp [hunk 1]
```diff
@@ -131,6 +131,15 @@ void ansi_c_internal_additions(std::string &code)
     // NOLINTNEXTLINE(whitespace/line_length)
     "__CPROVER_bool __CPROVER_threads_exited[__CPROVER_constant_infinity_uint];\n"
     "unsigned long __CPROVER_next_thread_id=0;\n"
+
+    // traces
+    "void CBMC_trace(int lvl, const char *event, ...);\n"
+
+    // pointers
+    "unsigned __CPROVER_POINTER_OBJECT(const void *p);\n"
```
Comment (smowton) [inline src/ansi-c/ansi_c_internal_additions.cpp]: Convention: I think these are usually `__CPROVER_lowercase_name`

**Comment by kroening** [comment]
BTW, there's now __CPROVER_ssize_t!

File: src/ansi-c/ansi_c_internal_additions.cpp (hunk 1 above)
Comment (kroening) [reply src/ansi-c/ansi_c_internal_additions.cpp]: The above seems to have been re-introduced after it got moved out into cprover_builtin_headers.h

File: src/ansi-c/library/string.c [hunk 2]
```diff
@@ -140,9 +142,16 @@ inline char *strcpy(char *dst, const char *src)
...
+    {
+    unsigned long n;
+    for(n = 0U; *(src + n) != 0; ++n)
+        ;
+    __CPROVER_precondition(
+      __CPROVER_POINTER_OBJECT(dst) != __CPROVER_POINTER_OBJECT(src) ||
+      __CPROVER_POINTER_OFFSET(src) + n < __CPROVER_POINTER_OFFSET(dst) ||
+      __CPROVER_POINTER_OFFSET(dst) + n < __CPROVER_POINTER_OFFSET(src),
+      "strcpy src/dst overlap");
+    }
```
Comment (kroening) [inline src/ansi-c/library/string.c:154]: The precondition needs to be at the head of the function!

**Comment by TGWDB** [comment]
This appears SV-COMP'18 specific and not planned to be applied to the whole codebase (also flagged to not be merged and still requiring author cleanup, regression tests, etc.).

If you believe this close is erroneous please reopen.

**Comment by martin-cs** [comment]
@tautschnig is this still relevant for the current SV-COMPs?

**Pull request closed without merging**